backend/fbdev.c

```diff
@@ -0,0 +1,182 @@
+/*
+ * Twin - A Tiny Window System
+ * Copyright (c) 2024 National Cheng Kung University, Taiwan
+ * All rights reserved.
+ */
+
+#include <fcntl.h>
+#include <linux/fb.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <sys/ioctl.h>
+#include <sys/mman.h>
+#include <twin.h>
+#include <unistd.h>
+
+#include "linux_input.h"
+#include "twin_backend.h"
+
+typedef struct {
+    int fd;
+    int depth;
+    int image_y;
+    int *pixels;
+    int *fbdev_raw;
+    twin_screen_t *screen;
+    void *input;
+} twin_fbdev_t;
+
+#define FBDEV_NAME "FRAMEBUFFER"
+#define FBDEV_DEFAULT "/dev/fb0"
+#define SCREEN(x) ((twin_context_t *) x)->screen
+#define PRIV(x) ((twin_fbdev_t *) ((twin_context_t *) x)->priv)
+
+static void _twin_fbdev_put_span(twin_coord_t left,
+                                 twin_coord_t top,
+                                 twin_coord_t right,
+                                 twin_argb32_t *pixels,
+                                 void *closure)
+{
+    twin_screen_t *screen = SCREEN(closure);
+    twin_fbdev_t *tx = PRIV(closure);
+
+    for (twin_coord_t ix = left, iy = top; ix < right; ix++) {
```

Comment:
Suppose that we extend `twin_fbdev_t` with the following members:
```c
    ...
    bool vt_active;  /* vt is currently active */
    bool active; /* screen is active and useable */

    /* fbdev setup */
    struct fb_var_screeninfo fb_var;
    struct fb_fix_screeninfo fb_fix;
    uint16_t cmap[3][256];
     uint8_t *fb_base; /* better replacement for fbdev_raw */
    size_t fb_len;
    uint8_t *fb_ptr;
} twin_fbdev_t;
```

We can prepare the setup function below:
```c
static bool twin_fbdev_apply_config(twin_fbdev_t *tx)
{
    off_t off, pgsize = getpagesize();
    struct fb_cmap cmap;
    size_t len;

    /* Tweak fields to default to 32 bpp argb and virtual == phys */
    tx->fb_var.xres_virtual = tx->fb_var.xres;
    tx->fb_var.yres_virtual = tx->fb_var.yres;
    tx->fb_var.bits_per_pixel = 32;

    /* Apply fbdev settings */
    if (ioctl(tx->fb_fd, FBIOPUT_VSCREENINFO, &tx->fb_var) < 0) {
        fprintf(stderr, "Failed to set framebuffer mode");
        return false;
    }

    /* Get new fbdev configuration */
    if (ioctl(tx->fb_fd, FBIOGET_VSCREENINFO, tx->fb_var) < 0) {
        fprintf(stderr, "Failed to get framebuffer information");
        return false;
    }

    /* Check bits per pixel */
    if (tx->fb_var.bits_per_pixel != 32) {
        fprintf(stderr, "Failed to set fb bpp to 32");
        return false;
    }

    /* Set colormap */
    cmap.start = 0;
    cmap.len = 256;
    cmap.red = tx->cmap[0];
    cmap.green = tx->cmap[1];
    cmap.blue = tx->cmap[2];
    cmap.transp = NULL;
    ioctl(tx->fb_fd, FBIOPUTCMAP, &cmap);

    /* Get remaining settings */
    ioctl(tx->fb_fd, FBIOGET_FSCREENINFO, &tx->fb_fix);
    /* Map the fb */
    off = (off_t) tx->fb_fix.smem_start & (pgsize - 1);
    len = (size_t) tx->fb_fix.smem_len + off + (pgsize - 1);
    len &= ~(pgsize - 1);
    tx->fb_len = len;

    tx->fb_base =
        mmap(NULL, len, PROT_READ | PROT_WRITE, MAP_SHARED, tx->fb_fd, 0);
    if (tx->fb_base == MAP_FAILED) {
        fprintf(stderr, "Failed to mmap framebuffer");
        return 0;
    }
    tx->fb_ptr = tx->fb_base + off;

    return true;
```

And mode switch function:
```
static void twin_fbdev_switch(twin_fbdev_t *tx, int activate)
{
    tx->vt_active = activate;

    /* Upon activation */
    if (activate) {
        /* Switch complete */
        ioctl(tx->vt_fd, VT_RELDISP, VT_ACKACQ);

        /* Restore fbdev settings */
        if (twin_fbdev_apply_config(tx)) {
            tx->active = true;

            /* Mark entire screen for refresh */
            if (tx->screen)
                twin_screen_damage(tx->screen, 0, 0, tx->screen->width,
                                   tx->screen->height);
        }
    } else {
        /* Allow switch. 
         * FIXME: expose some option to disallow the swith
         */
        ioctl(tx->vt_fd, VT_RELDISP, 1);

        tx->active = false;

        if (tx->fb_base != MAP_FAILED)
            munmap(tx->fb_base, tx->fb_len);
        tx->fb_base = MAP_FAILED;
    }
}

static int vt_switch_pending;

static bool twin_fbdev_work(void *closure)
{
    twin_fbdev_t *tx = closure;

    if (vt_switch_pending) {
        twin_fbdev_switch(tx, !tx->vt_active);
        vt_switch_pending = false;
    }

    if (tx->screen && tx->active && twin_screen_damaged(tx->screen))
        twin_screen_update(tx->screen);

    return true;
}

static void twin_fbdev_vtswitch(int sig)
{
    signal(sig, twin_fbdev_vtswitch);
    vt_switch_pending = true;
}

static bool twin_fbdev_setup_vt(twin_fbdev_t *tx, int switch_sig)
{
    ...
    signal(switch_sig, twin_fbdev_vtswitch);
    ...
}
```

Rewrite the function body as following:
```c
{
    twin_screen_t *screen = SCREEN(closure);
    twin_fbdev_t *tx = PRIV(closure);

    if (tx->fbdev_raw == MAP_FAILED)
        return;

    twin_coord_t    width = right - left;
    uint32_t *dest = (uint32_t *) (tf->fb_ptr + top * tf->fb_fix.line_length);
    dest += left;
    memcpy(dest, pixels, width * 4);
}
```